### PR TITLE
Enable Structured Variable Replacement for CloudFormation

### PR DESCRIPTION
**Pull request by iryan2**
This PR enables the Structured Variable Replacement feature for CloudFormation deployments.

The included test creates a Stack from a template, performs variable replacement for the name of an S3 bucket, and verifies that the bucket with the replaced name can be found. 

The test can be made to fail by removing the `s3-full-access` permission from the user in IAM.

### CODE REVIEW

**Comment by iryan2** [comment]
[CloudFormation: Enable structured variable replacements](https://trello.com/c/dTVHBReU/85-cloudformation-enable-structured-variable-replacements)

File: source/Calamari.Tests/AWS/CloudFormation/CloudFormationFixture.cs [hunk 1]
```diff
@@ -41,157 +17,37 @@ public class CloudFormationFixture
...
-        }}
-    }}
-}}";
+            var templateFilePath = CloudFormationFixtureHelpers.WriteTemplateFile(CloudFormationFixtureHelpers.GetBasicS3Template(StackName));
+            var variables = new CalamariVariables();
+
             try
             {
-                DeployTemplate(StackName, template);
+                CloudFormationFixtureHelpers.DeployTemplate(StackName, templateFilePath, variables);
```
Comment (geofflamrock) [inline source/Calamari.Tests/AWS/CloudFormation/CloudFormationFixture.cs:25]: You can prob just inline the instance of CalamariVariables here if you don't need to add anything specific into it (and remove changes on line 21)
```suggestion
                CloudFormationFixtureHelpers.DeployTemplate(StackName, templateFilePath, new CalamariVariables());
```

File: source/Calamari.Tests/AWS/CloudFormation/CloudFormationFixture.cs [hunk 2]
```diff
@@ -41,157 +17,37 @@ public class CloudFormationFixture
...
-        {
-            var variablesFile = Path.GetTempFileName();
-            var variables = new CalamariVariables();
-            variables.Set("Octopus.Action.AwsAccount.Variable", "AWSAccount");
-            variables.Set("AWSAccount.AccessKey", Environment.GetEnvironmentVariable("AWS_Calamari_Access"));
-            variables.Set("AWSAccount.SecretKey", Environment.GetEnvironmentVariable("AWS_Calamari_Secret"));
-            variables.Set("Octopus.Action.Aws.Region", RegionEndpoint.APSoutheast1.SystemName);
-            variables.Save(variablesFile);
-
+            var template = CloudFormationFixtureHelpers.GetBasicS3Template(StackName);
```
Comment (geofflamrock) [inline source/Calamari.Tests/AWS/CloudFormation/CloudFormationFixture.cs:41]: We could use the same helper methods to create the template as used in the above test here

File: source/Calamari.Tests/AWS/CloudFormation/CloudFormationFixtureHelpers.cs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+            var templateFilePath = Path.GetTempFileName();
+            File.WriteAllText(templateFilePath, template);
+            return templateFilePath;
+        }
+
+        public static async Task ValidateS3BucketExists(string stackName)
+        {
+            await ValidateS3(async client =>
+            {
+                // Bucket with replaced name was created successfully
```
Comment (geofflamrock) [inline source/Calamari.Tests/AWS/CloudFormation/CloudFormationFixtureHelpers.cs:50]: nit: The comment here looks like its out of date, its mentioning a replaced name

**Review by geofflamrock — COMMENTED** [review]
Looking pretty good, just had a few minor improvements that could be made

**Comment by iryan2** [comment]
@geofflamrock Thanks for the feedback and nitpicks! Should we merge this into your branch?

**Review by geofflamrock — APPROVED** [review]
LGTM, let's get it into the logging branch and then we can work on getting the tests running in CI along with those ones and get it in :)

**Pull request merged**